source/extensions/filters/network/thrift_proxy/protocol.h

```diff
@@ -0,0 +1,191 @@
+#pragma once
+
+#include <memory>
+#include <stack>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/common/pure.h"
+
+#include "common/common/assert.h"
+#include "common/common/macros.h"
+#include "common/singleton/const_singleton.h"
+
+#include "absl/types/optional.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace ThriftProxy {
+
+/**
+ * Names of available Protocol implementations.
+ */
+class ProtocolNameValues {
```

Comment:
It looks like the only use of this class is for each `Protocol` subclass to return a string representation of its own name. I think it would be cleaner to just make each `Protocol` subclass define its own name string internally. That way, we won't have to modify this central `ProtocolNameValues` class every time a new `Protocol` subclass is added.

Reply:
By and large these are used only during debug logging, but my original intention was to avoid allocating these strings. At some point I decided to make the auto protocol report its underlying protocol and ended up constructing or copying strings in all the name() methods. 

Since I later added `setProtocol` to make the AutoProtocolImpl easier to test, I went ahead and made AutoProtocolImpl::name() return a field value, updated by setProtocol. This allowed me to have Protocol::name() return a `const std::string&`. So no more allocations, but I still have ProtocolNames.

I can move these strings into the Protocol implementations, but per the style guide, I'd need to use the CONSTRUCT_ON_FIRST_USE macro. AutoProtocolImpl would require it's own private method to wrap CONSTRUCT_ON_FIRST_USE for "auto". I *think* putting all these strings in this class and using ConstSingleton ends up being a little cleaner.